edgelet/hsm-sys/azure-iot-hsm-c/src/edge_hsm_client_store.c

```diff
@@ -1692,40 +1705,171 @@ static int generate_edge_hsm_certificates_if_needed(void)
+    *device_id_pk_path = pk_path;
+
+    return result;
+}
+
+static int hsm_provision_edge_id_certificate(void)
+{
+    int result;
+    char *device_id_cert_path = NULL;
+    char *device_id_pk_path = NULL;
+    bool env_set = false;
+    unsigned int mask = 0, i = 0;
+
+    if (get_device_id_env_vars(&device_id_cert_path, &device_id_pk_path) != 0)
+    {
+        result = __FAILURE__;
+    }
+    else
+    {
+        if (device_id_cert_path != NULL)
+        {
+            if ((strlen(device_id_cert_path) != 0) && is_file_valid(device_id_cert_path))
+            {
+                mask |= 1 << i; i++;
+            }
+            else
+            {
+                LOG_ERROR("Path set in env variable %s is invalid or cannot be accessed: '%s'",
+                          ENV_DEVICE_ID_CERTIFICATE_PATH, device_id_cert_path);
+
+            }
+            env_set = true;
+            LOG_DEBUG("Env %s set to %s", ENV_DEVICE_ID_CERTIFICATE_PATH, device_id_cert_path);
+        }
+        else
+        {
+            LOG_DEBUG("Env %s is NULL", ENV_DEVICE_ID_CERTIFICATE_PATH);
+        }
+
+        if (device_id_pk_path != NULL)
+        {
+            if ((strlen(device_id_pk_path) != 0) && is_file_valid(device_id_pk_path))
+            {
+                mask |= 1 << i; i++;
+            }
+            else
+            {
+                LOG_ERROR("Path set in env variable %s is invalid or cannot be accessed: '%s'",
+                          ENV_DEVICE_ID_PRIVATE_KEY_PATH, device_id_pk_path);
+
+            }
+            env_set = true;
+            LOG_DEBUG("Env %s set to %s", ENV_DEVICE_ID_PRIVATE_KEY_PATH, device_id_pk_path);
+        }
+        else
+        {
+            LOG_DEBUG("Env %s is NULL", ENV_DEVICE_ID_PRIVATE_KEY_PATH);
+        }
+
+        LOG_DEBUG("Device identity certificate setup mask 0x%02x", mask);
+
+        if (env_set && (mask != 0x3))
+        {
+            LOG_ERROR("To setup the Edge device certificates, set "
+                      "env variables with valid values:\n  %s\n  %s",
+                      ENV_DEVICE_ID_CERTIFICATE_PATH, ENV_DEVICE_ID_PRIVATE_KEY_PATH);
+            result = __FAILURE__;
+        }
+        // since we don't know the issuer, we treat the device certificate as the issuer
+        else if (env_set && (edge_hsm_client_store_insert_pki_cert(g_crypto_store,
+                                                                   EDGE_DEVICE_ALIAS,
+                                                                   EDGE_DEVICE_ALIAS,
+                                                                   device_id_cert_path,
+                                                                   device_id_pk_path) != 0))
+        {
+            LOG_ERROR("Failure inserting device identity certificate and key into the HSM store");
+            result = __FAILURE__;
+        }
+        else
+        {
+            result = 0;
+        }
+    }
+
+    if (device_id_cert_path != NULL)
```

Comment:
If you move `FREEIF` to a common header you an use it here. (Though like my comment on the macro, the `!= NULL` test isn't needed for `free` anyway.

Reply:
Done.

Reply:
I guess you don't want to use `FREEIF` here since they don't need to be set to `NULL`? In that case, like I said in the comment, just remove the `!= NULL` test.